### PR TITLE
Add `set_empty_query_match_all` method

### PR DESCRIPTION
**Pull request by druzn3k**
This commit adds the ability to change the default behavior of an empty
search query, from returning `QueryEmpty` to returning `AllQuery`,
matching every value in the storage.

Closes issue #386

### CODE REVIEW

File: src/query/query_parser/query_parser.rs [hunk 1]
```diff
@@ -181,6 +202,10 @@ impl QueryParser {
     /// Implementing a lenient mode for this query parser is tracked
     /// in [Issue 5](https://github.com/fulmicoton/tantivy/issues/5)
     pub fn parse_query(&self, query: &str) -> Result<Box<Query>, QueryParserError> {
+        if query.trim().is_empty() && self.empty_query_match_all {
```
Comment (fulmicoton) [inline src/query/query_parser/query_parser.rs]: Please work of the userinputast -> logicalinputast conversion. Not on the string itself.

The analyzer might transform a query string into an empty query even when `query.trim().is_empty()`.
For instance, if the query only contains punctuation or stop words.

File: src/query/query_parser/query_parser.rs [hunk 2]
```diff
@@ -591,6 +616,19 @@ mod test {
         assert_eq!(format!("{:?}", query), "EmptyQuery");
     }
 
+    #[test]
+    pub fn test_parse_query_empty_match_all() {
+        test_parse_query_to_logical_ast_helper("", "<emptyclause>", false);
+        test_parse_query_to_logical_ast_helper(" ", "<emptyclause>", false);
+        let mut query_parser = make_query_parser();
```
Comment (fulmicoton) [inline src/query/query_parser/query_parser.rs:630]: Why wasn't it possible to work off the logical ast to test the match all case too?

File: src/query/query_parser/query_parser.rs (hunk 1 above)
Comment (druzn3k) [reply src/query/query_parser/query_parser.rs]: Ok, no problem. I tried first with the naive solution to see if I could do it :)

File: src/query/query_parser/query_parser.rs (hunk 2 above)
Comment (druzn3k) [reply src/query/query_parser/query_parser.rs:630]: Lack of confidence with the language mostly. I tried first with a "good enough" solution to see if i could do it. Now I'll try to improve on it. Thanks for the patience

**Comment by druzn3k** [comment]
I updated the code, and I want to punch myself in the face for not having seen the solution before. Oh well, the struggle is half the fun. Let me know if this is good or needs another iteration. Thanks! :+1:

File: src/query/query_parser/query_parser.rs [hunk 3]
```diff
@@ -456,12 +468,12 @@ fn convert_literal_to_query(logical_literal: LogicalLiteral) -> Box<Query> {
     }
 }
 
-fn convert_to_query(logical_ast: LogicalAST) -> Box<Query> {
+fn convert_to_query(logical_ast: LogicalAST, should_match_all: bool) -> Box<Query> {
```
Comment (fulmicoton) [inline src/query/query_parser/query_parser.rs:471]: `should_match_all` should be renamed `should_match_all_if_empty`.

File: src/query/query_parser/query_parser.rs [hunk 4]
```diff
@@ -456,12 +468,12 @@ fn convert_literal_to_query(logical_literal: LogicalLiteral) -> Box<Query> {
...
     match trim_ast(logical_ast) {
         Some(LogicalAST::Clause(trimmed_clause)) => {
             let occur_subqueries = trimmed_clause
                 .into_iter()
-                .map(|(occur, subquery)| (occur, convert_to_query(subquery)))
+                .map(|(occur, subquery)| (occur, convert_to_query(subquery, should_match_all)))
```
Comment (fulmicoton) [inline src/query/query_parser/query_parser.rs:476]: I don't think this is correct to recursively apply this value.

What happens if we have queries like :
`A AND NOT <some_stopword>`

This "what happens if empty logic should on" if the entire query is empty

File: src/query/query_parser/query_parser.rs (hunk 2 above)
Comment (fulmicoton) [reply src/query/query_parser/query_parser.rs:630]: This has not been fixed.

Comment (druzn3k) [reply src/query/query_parser/query_parser.rs:630]: sorry I forgot! :(

File: src/query/query_parser/query_parser.rs (hunk 3 above)
Comment (druzn3k) [reply src/query/query_parser/query_parser.rs:471]: No problem

File: src/query/query_parser/query_parser.rs (hunk 4 above)
Comment (druzn3k) [reply src/query/query_parser/query_parser.rs:476]: Ok, I'll try to find a more correct way.

Comment (fulmicoton) [reply src/query/query_parser/query_parser.rs:476]: To be honest I am not entirely sure of what the best definition should be. You just write a bunch of unit test beforehands, and we can discuss the best outcome when `should_match_all` is `true` or `false`.

- `something -<a stop word>`
- `something something_else`
- `something AND something_else`
- `something AND NOT something_else`
- `something OR something_else`

Comment (druzn3k) [reply src/query/query_parser/query_parser.rs:476]: So, I wrote this at the moment:

```rust
pub fn test_logic_behind_should_match_all() {
    let query_parser = make_query_parser();

    // something -<a stop word>
    assert!(query_parser.parse_query("toto -such").is_ok());

    // something something_else
    assert!(query_parser.parse_query("toto teta").is_ok());

    // something AND something_else
    assert!(query_parser.parse_query("toto AND foo").is_ok());

    // something AND NOT something_else
    assert!(query_parser.parse_query("toto AND NOT bar").is_ok());

    // something OR something_else
    assert!(query_parser.parse_query("toto OR foobar").is_ok());
}
```

I plan to make the `query_parser` variable `mut` and toggle the `should_match_all` on/off on need. It's ok, or should I make a new `query_parser` from scratch for every test case? The queries are correct or did you expect something different? Sorry for not reaching out before, I got kind of busy :)

Comment (fulmicoton) [reply src/query/query_parser/query_parser.rs:476]: @druzn3k 

Sorry for the 3month iatus. Of course all of these queries should pass.
My concern is more about what they should return...

For instance, if empty match all,  `term_that_exists -term_that_does_not_exist`  will return no document, which is rather unexpected.

I think the match_all_if_query should not apply if this is part of another query.

**Pull request closed without merging**